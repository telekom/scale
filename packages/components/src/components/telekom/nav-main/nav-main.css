/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

scale-nav-main {
  --spacing-x: var(--telekom-typography-font-size-body);
  --color: var(--telekom-color-text-and-icon-standard);
  --color-hover: var(--telekom-color-text-and-icon-primary-hovered);
  --color-active: var(--telekom-color-text-and-icon-primary-hovered);
  --color-selected: var(--telekom-color-text-and-icon-primary-standard);
  --font-size: var(--telekom-typography-font-size-callout);
  --font-weight: var(--telekom-typography-font-weight-extra-bold);
  --line-height: var(--telekom-typography-line-spacing-tight);
}

.main-navigation__item {
  height: 100%;
  padding-left: var(--spacing-x);
  padding-right: var(--spacing-x);
}
.main-navigation__item:has(a[href]) {
  cursor: pointer;
}
.main-navigation__item-link-text {
  white-space: nowrap;
}

.main-navigation__item.mega-menu--visible .mega-menu {
  display: block;
}
.main-navigation__item.mega-menu--visible .mega-menu__wrapper {
  pointer-events: visible;
}
.main-navigation__item .main-navigation__item-link {
  display: flex;
  height: 100%;
  color: var(--color);
  font-size: var(--font-size);
  font-weight: var(--font-weight);
  line-height: var(--line-height);
  transition: color, border 0.2s ease-in-out;
  align-items: center;
  text-decoration: none;
}
.main-navigation__item:hover .main-navigation__item-link {
  color: var(--color-hover);
}
.main-navigation__item:hover .main-navigation__item-link-text {
  border-bottom: 2px solid var(--color-hover);
  color: var(--color-hover);
  margin-bottom: -2px;
  height: calc(100% - 4px);
  display: flex;
  align-items: center;
}
.main-navigation__item:active .main-navigation__item-link {
  color: var(--color-active);
}
.main-navigation__item:active .main-navigation__item-link-text {
  border-bottom: 2px solid var(--color-active);
  margin-bottom: -2px;
  height: calc(100% - 4px);
  display: flex;
  align-items: center;
}
.main-navigation__item.selected .main-navigation__item-link {
  color: var(--color-selected);
}

.main-navigation__item.selected:hover .main-navigation__item-link {
  color: var(--color-selected);
}
.main-navigation__item.selected .main-navigation__item-link-text {
  border-bottom: 2px solid var(--color-selected);
  margin-bottom: -2px;
  height: calc(100% - 4px);
  display: flex;
  align-items: center;
}
.main-navigation__item.selected:hover .main-navigation__item-link-text {
  color: var(--color-selected);
  border-bottom: 2px solid var(--color-selected);
}
.main-navigation__item.mega-menu--visible .main-navigation__item-link-text {
  border-bottom: 2px solid var(--color-hover);
  margin-bottom: -2px;
  height: calc(100% - 4px);
  display: flex;
  align-items: center;
}
.sr-only {
  position: absolute;
  left: -10000px;
  overflow: hidden;
}
