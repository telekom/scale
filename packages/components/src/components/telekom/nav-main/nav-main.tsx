/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { Component, Prop, h, Element, Host } from '@stencil/core';
import classNames from 'classnames';
import { HTMLStencilElement } from '@stencil/core/internal';

@Component({
  tag: 'scale-nav-main',
  styleUrl: './nav-main.css',
})
export class NavMain {
  @Element() hostElement: HTMLStencilElement;
  /** (optional) if this item is active */
  // DEPRECATED - active should replace isActive
  @Prop() isActive: boolean;
  @Prop() active: boolean;
  /** (optional) if this mega-menu is visible */
  @Prop() isMegaMenuVisible?: boolean = false;
  /** (optional) href value */
  @Prop() href?: string = 'javascript:void(0);';
  /** (optional) name value */
  @Prop() name: string;

  @Prop() clickLink: any;
  hasSlotMegaMenu: boolean;

  componentWillLoad() {
    this.hasSlotMegaMenu = !!this.hostElement.querySelector('app-mega-menu');
  }
  render() {
    return (
      <Host>
        <li class={this.getCssClassMap()}>
          <a
            class="main-navigation__item-link"
            href={this.href}
            aria-current={this.active || this.isActive ? 'true' : 'false'}
            aria-haspopup={this.hasSlotMegaMenu ? 'true' : 'false'}
            tabIndex={0}
            onClick={this.clickLink}
          >
            <span class="main-navigation__item-link-text">{this.name}</span>
            {(this.active || this.isActive) && (
              <span class="sr-only">active</span>
            )}
          </a>
          <slot></slot>
        </li>
      </Host>
    );
  }

  getCssClassMap() {
    return classNames(
      'main-navigation__item',
      this.isMegaMenuVisible && 'mega-menu--visible',
      (this.active || this.isActive) && 'selected'
    );
  }
}
