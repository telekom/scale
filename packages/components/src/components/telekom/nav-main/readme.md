# scale-nav-main



<!-- Auto Generated Below -->


## Properties

| Property            | Attribute              | Description                             | Type      | Default                 |
| ------------------- | ---------------------- | --------------------------------------- | --------- | ----------------------- |
| `active`            | `active`               |                                         | `boolean` | `undefined`             |
| `clickLink`         | `click-link`           |                                         | `any`     | `undefined`             |
| `href`              | `href`                 | (optional) href value                   | `string`  | `'javascript:void(0);'` |
| `isActive`          | `is-active`            | (optional) if this item is active       | `boolean` | `undefined`             |
| `isMegaMenuVisible` | `is-mega-menu-visible` | (optional) if this mega-menu is visible | `boolean` | `false`                 |
| `name`              | `name`                 | (optional) name value                   | `string`  | `undefined`             |
| `popup`             | `popup`                |                                         | `boolean` | `undefined`             |


## Dependencies

### Used by

 - [scale-app-header](../app-header)

### Graph
```mermaid
graph TD;
  scale-app-header --> scale-nav-main
  style scale-nav-main fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
