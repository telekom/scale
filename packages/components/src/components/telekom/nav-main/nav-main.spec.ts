/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newSpecPage } from '@stencil/core/testing';
import { NavMain } from './nav-main';

describe('NavMain', () => {
  let page;
  beforeEach(async () => {
    page = await newSpecPage({
      components: [NavMain],
      html: `<scale-nav-main></scale-nav-main>`,
    });
  });
  it('should match snapshot', async () => {
    expect(page.root).toMatchSnapshot();
  });
  it('should handle is active props', async () => {
    const specPage = await newSpecPage({
      components: [NavMain],
      html: `
      <scale-nav-main
        is-active=true
        is-mega-menu-visible=true
        name='nameDummy'
      ></scale-nav-main>`,
    });
    expect(specPage.root).toMatchSnapshot();
  });
  it('should handle css classes', () => {
    const element = new NavMain();

    element.isMegaMenuVisible = true;
    expect(element.getCssClassMap()).toContain('mega-menu--visible');

    // DEPRECATED - active should replace isActive
    element.isActive = true;
    expect(element.getCssClassMap()).toContain('selected');

    element.active = true;
    expect(element.getCssClassMap()).toContain('selected');
  });
});
