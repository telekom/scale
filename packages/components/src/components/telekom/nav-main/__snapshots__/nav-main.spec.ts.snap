// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavMain should handle is active props 1`] = `
<scale-nav-main is-active="true" is-mega-menu-visible="true" name="nameDummy">
  <!---->
  <li class="main-navigation__item mega-menu--visible selected">
    <a aria-current="true" aria-haspopup="false" class="main-navigation__item-link" href="javascript:void(0);">
      <span class="main-navigation__item-link-text">
        nameDummy
      </span>
      <span class="sr-only">
        active
      </span>
    </a>
  </li>
</scale-nav-main>
`;

exports[`NavMain should match snapshot 1`] = `
<scale-nav-main>
  <!---->
  <li class="main-navigation__item">
    <a aria-current="false" aria-haspopup="false" class="main-navigation__item-link" href="javascript:void(0);">
      <span class="main-navigation__item-link-text"></span>
    </a>
  </li>
</scale-nav-main>
`;
